File: 5.Linked Lists/python/singly-linked-list/SinglyLinkedList.py
class SinglyLinkedList:
    """ The implementation of Singly Linked List ADT """

    class _Node:
        """ Helper class for singly linked list storage """
        __slots__ = 'element', 'next'

        def __init__(self, element, next_node=None):
            self.element = element
            self.next = next_node

    def __init__(self):
        """ Creates an empty list """
        self._head = None
        self._tail = None
        self._size = 0

    def __len__(self):
        """ Returns the number of elements of the list """
        return self._size

    def __iter__(self):
        """ Makes the list iterable """
        pass

    def is_empty(self):
        """ Returns true if the list is empty """
        return self._size == 0

    def add_front(self, element):
        """ Adds an element to the front of the list """
        self._head = self._Node(element, self._head)
        self._size += 1
        if self._size == 1:
            self._tail = self._head

    def add_last(self, element):
        """ Adds an element to the end of the list """
        new_node = self._Node(element)
        if self._size == 0:
            self._head = new_node
        else:
            self._tail.next = new_node
        self._tail = new_node
        self._size += 1

    def remove_front(self):
        """ Removes an element from the front of the list """
        if self._size == 0:
            raise Empty('List is empty')

        self._head = self._head.next
        self._size -= 1

        if self._size == 0:
            self._tail = None

    def remove_last(self):
        """ Removes an element from the end of the list """
        if self._size == 0:
            raise Empty('List is empty')

        if self._size == 1:
            self._head = None
            self._tail = None
        else:
            node = self._head
            while node.next != self._tail:
                node = node.next
            node.next = None
            self._tail = node
        self._size -= 1



class Empty(Exception):
    """ Error attempting to access an element from an empty list """
    pass


def main():

    sll = SinglyLinkedList()
    sll.add_front('A')
    # sll.add_front('B')
    # sll.add_last('Y')
    # sll.add_last('Z')

    sll.remove_front()

    # A, B, C, D
    print(len(sll))
    print(sll._head)
    print(sll._tail)

if __name__ == '__main__':
    main()

File: 5.Linked Lists/javascript/singly-linked-list/SinglyLinkedList.js
const Node = require('./Node');

class SinglyLinkedList {
  constructor() {
    // Contains node reference as head
    this._head = null;

    // Contains node reference as tail
    this._tail = null;

    // Determines size of the list
    this._size = 0;
  }

  /**
   * Get the current head
   * 
   * @return {Node}
   */
  getHead() {
    return this._head;
  }

  /**
   * Get the current tail
   */
  getTail() {
    return this._tail;
  }

  /**
   * Gets the list size
   * 
   * @param {Integer} data 
   */
  getSize() {
    return this._size;
  }

  /**
   * Checks if the list is empty
   * 
   * @param {Boolean} data 
   */
  isEmpty() {
    return this._size === 0;
  }

  /**
   * Adds an item to the front of the list
   * 
   * @param {String|Integer} data
   * @return void 
   */
  addFirst(data) {
    const node = new Node(data);
    const temp = this._head;

    this._head = node;
    this._head.setNext(temp);

    this._size++;

    if (this._size == 1) {
      this._tail = this._head;
    }
  }

  /**
   * Adds an item to the end of the list
   * 
   * @param {String|Integer} data 
   * @return void
   */
  addLast(data) {
    const node = new Node(data);
    if (this._size == 0) {
      this._head = node;
    } else {
      this._tail.setNext(node);
    }

    this._tail = node;
    this._size++;
  }

  /**
   * Removes an item from the front of the list 
   * 
   * @return void
   */
  removeFirst() {
    if (this._size !== 0) {
      this._head = this._head.getNext();
      this._size--;
    }

    if (this._size === 0) {
      this._tail = null;
    }
  }

  /**
   * Removes an item from the end of the list
   * 
   * @return void
   */
  removeLast() {
    if (this._size !== 0) {
      if (this._size == 1) {
        this._head = null;
        this._tail = null;
      } else {
        let current = this._head;
        while (current.getNext() !== this._tail) {
          current = current.getNext();
        }

        current.setNext(null);
        this._tail = current;
      }

      this._size--;
    }
  }

  /**
   * Finds an item from the list
   * 
   * @param {String|Integer} data
   * @return {Node|Boolean} 
   */
  find(data) {
    if (this.isEmpty()) {
      return false;
    }

    let node = this._head;
    while (node !== null) {
      if (node.getData() === data) {
        return node;
      }

      node = node.getNext();
    }

    return false;
  }

  /**
   * Insert an item in the list after the specified item
   * 
   * @param {String|Integer} insert 
   * @param {String|Integer} after 
   * @return {Node|Boolean}
   */
  insert(insertThisElement, afterThisElement) {
    if (this.isEmpty()) {
      return false;
    }

    let node = this._head;
    while (node !== null) {
      if (node.getData() === afterThisElement) {
        const newNode = new Node(insertThisElement);

        if (node.getNext() == null) {
          this._tail = newNode;
        }

        newNode.setNext(node.getNext());
        node.setNext(newNode);

        this._size++;

        return newNode;
      }

      node = node.getNext();
    }

    return false;
  }

  display() {
    if (this.isEmpty()) {
      return null;
    }

    let node = this._head;
    let items = [];
    while (node !== null) {
      items.push(node.getData());
      node = node.getNext();
    }

    return items;
  }
}

module.exports = SinglyLinkedList;